test/main.cpp

```diff
@@ -231,6 +232,23 @@ static bool test14()
     }
 }
 
+static bool test15()
+{
+    // Avoid cumulated error
+    PolylineEncoder encoder;
+    encoder.addPoint(47.231174468994141, 16.62629508972168);
```

Comment:
Aha. Your test is not "clean". As a result your output differ from what Google page shows. It's clear that these values are not the same Google script uses for calculations on the web page. The thing is that you cannot even enter these values in web form. As you probably noticed it rounds the input up to five places after the decimal point. Thus `0.0000005` becomes `0.0` and `47.231174468994141` becomes `47.23117`. Of course it will give you different results. Thus, if you want to conform with the Google results you should rather perform the same rounding for all points. I think, the right fix will be as follows:

````diff
diff --git a/src/polylineencoder.cpp b/src/polylineencoder.cpp
index e53c983..603d08f 100644
--- a/src/polylineencoder.cpp
+++ b/src/polylineencoder.cpp
@@ -39,8 +39,9 @@ void PolylineEncoder::addPoint(double latitude, double longitude)
 {
     assert(latitude <= 90.0 && latitude >= -90.0);
     assert(longitude <= 180.0 && longitude >= -180.0);
-
-    m_polyline.emplace_back(latitude, longitude);
+
+    m_polyline.emplace_back(std::round(latitude * s_presision) / s_presision,
+                            std::round(longitude * s_presision) / s_presision);
 }

 std::string PolylineEncoder::encode() const
````
This patch handles the test14 as well, btw.

Reply:
This could be a solution, if the only way of use is the addPoint() and then encode. However there is no guarantee that no-one would use the lib with the encode(polyline) function, and in that scenario, the polyline encoder fails to satisfy it's mission (technically the result would differ from the expected).
My concept is to keep the precision as long as possible and truncate it at the very last point when needed.

Reply:
> This could be a solution, if the only way of use is the addPoint() and then encode. However there is no guarantee that no-one would use the lib with the encode(polyline) function, and in that scenario, the polyline encoder fails to satisfy it's mission (technically the result would differ from the expected).

You can get the expected results only if preconditions are satisfied, i.e. if the input has up to five places after decimal point. If this precondition is violated - it's not the library's defect, but caller's. That's why referred web site prevents arbitrary precision on the GUI level.
> My concept is to keep the precision as long as possible and truncate it at the very last point when needed.

What do you achieve with postponing it?

Reply:
> If this precondition is violated - it's not the library's defect, but caller's.

Since it's nowhere documented, but expected, it's the lib's defect. And the referred page's GUI prevents arbitrary precision to satisfy the lib's inner expectation. Since you cannot guarantee the lib's expectations, and it's a common usage to try to encode real world geo coordinates, it should handled.

> What do you achieve with postponing it?

The reasoning is the previous

Reply:
TL; DR: if you don't want this fix in your lib, then drop the PR and i'll continue developing it in my own fork

Reply:
That's true. It's not documented, and this issue should be handled by myself. You can, if you want, to file a ticket for this. I also agree, that this code should conform the algorithm and the results on the mentioned Google's web page.

Reply:
> I also agree, that this code should conform the algorithm and the results on the mentioned Google's web page.

If you agree, then why not use the provided fix? :)

Reply:
> TL; DR: if you don't want this fix in your lib, then drop the PR and i'll continue developing it in my own fork

I will try to fix it, indeed. Regarding your PR, I like the idea of having cmake support. If you want, you can leave cmake related change, though.

Reply:
> > I also agree, that this code should conform the algorithm and the results on the mentioned Google's web page.
> 
> If you agree, then why not use the provided fix? :)

Because I don't like it:) It's too invasive, changes semantics, it harder to understand the intention as the part of encoding algorithm moved to another function etc.

Reply:
I propose:
- Add your tests, as they revealed the problem and are, therefore, valuable,
- Add your cmake support (you might add `polylineencoder.h` to the CMakeLists.txt too),
- Drop your `PolylineEncoder` fix, or use one I proposed (the last one).

I will do the rest.

Reply:
okay, i have an other idea, a non-intrusive one, stay tuned...